File: android/app/src/main/java/com/clovis/app/dossiers/DossiersDesignesRepository.kt
// Porte depuis clovis-mobile/android-legacy-natif (Lot 2 Partie 3, 23/08/2026)
// dans le plugin Capacitor (Lot 3B, 25/08/2026). Logique inchangee, seul le
// package a change (com.clovis.app.data -> com.clovis.app.dossiers).
//
// Storage Access Framework (SAF). L'etudiant designe un dossier une seule
// fois via le selecteur systeme (ACTION_OPEN_DOCUMENT_TREE) ; on prend une
// permission URI PERSISTANTE (takePersistableUriPermission) pour ne plus
// jamais avoir a la redemander, voir 02-fichiers-dossiers.md, "Objectif".
//
// Liste des dossiers designes stockee en local (SharedPreferences, Set<String>
// d'URI serialisees) : pas besoin de synchro serveur pour ca, c'est propre a
// cet appareil (l'acces SAF lui-meme est lie a l'appareil, pas au compte).
//
// Operations CRUD via DocumentFile (creer/lire/modifier) + DocumentsContract
// pour deplacer un document a l'interieur d'un meme arbre (moveDocument,
// dispo depuis l'API 24, minSdk du projet = 26).
package com.clovis.app.dossiers

import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.DocumentsContract
import android.util.Base64
import androidx.documentfile.provider.DocumentFile

private const val PREFS_NOM = "dossiers_designes"
private const val CLE_URIS = "uris"

data class DossierDesigne(val uri: Uri, val nom: String)

data class ElementDossier(
    val uri: Uri,
    val nom: String,
    val estDossier: Boolean,
    val tailleOctets: Long
)

// Ajoute le 30/08/2026 (correctif Claude chat, branchement lire_fichier /
// chercher_par_contenu, voir core/exploration_dossier_mobile.py cote
// clovis-backend et 04-lecture-contenu.md / 05-recherche-contenu-app-fermee.md
// a la racine de ce depot) : contenu brut d'un fichier, encode en base64,
// tel qu'attendu par core/lecture_fichier_mobile.py (qui applique lui-meme
// les seuils de taille par type et decide du traitement, rien a filtrer ici).
data class ContenuFichier(
    val contenuBase64: String,
    val typeMime: String,
    val nomFichier: String,
    val tailleOctets: Long
)

// Ajoute le 04/09/2026 (vectorisation en masse des dossiers designes,
// voir clovis-backend/api/dossiers_designes.py) : un fichier trouve en
// parcourant recursivement un dossier designe, avec son chemin de
// sous-dossiers depuis la racine (JAMAIS le nom du dossier designe
// lui-meme, ni le nom du fichier -- voir listerRecursif ci-dessous).
data class FichierAVectoriser(
    val uri: Uri,
    val nom: String,
    val chemin: List<String>,
    val typeMime: String,
    val tailleOctets: Long
)

class DossiersDesignesRepository(private val context: Context) {

    private val prefs = context.getSharedPreferences(PREFS_NOM, Context.MODE_PRIVATE)

    /** Dossiers designes actuellement, avec leur nom lisible. */
    fun listerDossiersDesignes(): List<DossierDesigne> {
        val uris = prefs.getStringSet(CLE_URIS, emptySet()) ?: emptySet()
        return uris.mapNotNull { brut ->
            val uri = Uri.parse(brut)
            val doc = DocumentFile.fromTreeUri(context, uri)
            if (doc != null && doc.exists()) {
                DossierDesigne(uri, doc.name ?: uri.lastPathSegment ?: "Dossier")
            } else {
                null // dossier devenu inaccessible (supprime/deplace hors de l'app), ignore silencieusement ici, nettoye au prochain ajout/retrait
            }
        }
    }

    /**
     * A appeler avec l'Uri renvoyee par le selecteur systeme
     * (ActivityResultContracts.OpenDocumentTree), apres que l'utilisateur a
     * choisi un dossier. Prend la permission persistante et l'ajoute a la liste.
     */
    fun ajouterDossierDesigne(uri: Uri) {
        context.contentResolver.takePersistableUriPermission(
            uri,
            Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
        )
        val actuels = (prefs.getStringSet(CLE_URIS, emptySet()) ?: emptySet()).toMutableSet()
        actuels.add(uri.toString())
        prefs.edit().putStringSet(CLE_URIS, actuels).apply()
    }

    /** Retire un dossier de la liste designee et libere la permission persistante. */
    fun retirerDossierDesigne(uri: Uri) {
        try {
            context.contentResolver.releasePersistableUriPermission(
                uri,
                Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
            )
        } catch (e: SecurityException) {
            // Permission deja perdue (dossier supprime cote systeme), rien a liberer.
        }
        val actuels = (prefs.getStringSet(CLE_URIS, emptySet()) ?: emptySet()).toMutableSet()
        actuels.remove(uri.toString())
        prefs.edit().putStringSet(CLE_URIS, actuels).apply()
    }

    /** Contenu direct d'un dossier (designe ou sous-dossier), trie dossiers puis fichiers, alphabetique. */
    fun listerContenu(dossierUri: Uri): List<ElementDossier> {
        val doc = DocumentFile.fromTreeUri(context, dossierUri) ?: DocumentFile.fromSingleUri(context, dossierUri)
        val enfants = doc?.listFiles() ?: emptyArray()
        return enfants
            .filter { it.name != null }
            .map { ElementDossier(it.uri, it.name!!, it.isDirectory, if (it.isDirectory) 0L else it.length()) }
            .sortedWith(compareBy({ !it.estDossier }, { it.nom.lowercase() }))
    }

    fun creerSousDossier(parentUri: Uri, nom: String): Boolean {
        val parent = DocumentFile.fromTreeUri(context, parentUri) ?: return false
        return parent.createDirectory(nom) != null
    }

    fun creerFichier(parentUri: Uri, nom: String, typeMime: String = "text/plain"): Boolean {
        val parent = DocumentFile.fromTreeUri(context, parentUri) ?: return false
        return parent.createFile(typeMime, nom) != null
    }

    /**
     * Lit le contenu brut d'un fichier deja repere (via listerContenu ou
     * une recherche), encode en base64 pour transiter par le canal temps
     * reel WebSocket (JSON, pas de binaire brut). Renvoie null si le
     * fichier est introuvable/illisible (deplace/supprime entre-temps,
     * permission perdue) : l'appelant traduit ca en erreur explicite,
     * jamais de silence.
     */
    fun lireFichier(elementUri: Uri): ContenuFichier? {
        val doc = DocumentFile.fromSingleUri(context, elementUri) ?: return null
        if (!doc.exists() || !doc.isFile) return null
        val nom = doc.name ?: return null
        val typeMime = doc.type ?: "application/octet-stream"
        val octets = lireOctets(elementUri) ?: return null
        return ContenuFichier(
            contenuBase64 = Base64.encodeToString(octets, Base64.NO_WRAP),
            typeMime = typeMime,
            nomFichier = nom,
            tailleOctets = octets.size.toLong()
        )
    }

    // Ajoute le 04/09/2026 : contenu brut d'un fichier (pas de base64,
    // contrairement a lireFichier ci-dessus qui l'encode pour le canal
    // temps reel WebSocket) -- utilise par la vectorisation en masse, qui
    // envoie les octets tels quels en multipart HTTP. lireFichier
    // reutilise maintenant cette meme fonction (aucune logique dupliquee).
    fun lireOctets(elementUri: Uri): ByteArray? {
        return try {
            context.contentResolver.openInputStream(elementUri)?.use { it.readBytes() }
        } catch (e: Exception) {
            null
        }
    }

    // Ajoute le 04/09/2026, Bourama : "apres avoir choisi un dossier, tout
    // ce qu'il contient hormis video est vectorise" -- parcourt
    // recursivement TOUS les sous-dossiers d'un dossier designe et
    // renvoie chaque FICHIER trouve (jamais les dossiers eux-memes), en
    // excluant la video (trop couteuse a vectoriser, decision explicite
    // de Bourama). `chemin` porte la liste ordonnee des noms de
    // sous-dossiers traverses pour atteindre ce fichier, PAS le nom du
    // dossier designe racine ni celui du fichier -- voir
    // clovis-backend/migrations/2026_09_04_dossiers_designes_vectorisation.sql.
    // Reutilise directement les DocumentFile enfants renvoyes par
    // listFiles() pour recurser (pas besoin de refaire fromTreeUri a
    // chaque niveau, l'URI enfant est deja valide dans l'arbre accorde).
    fun listerRecursif(dossierUri: Uri): List<FichierAVectoriser> {
        val racine = DocumentFile.fromTreeUri(context, dossierUri) ?: return emptyList()
        return parcourirRecursivement(racine, emptyList())
    }

    private fun parcourirRecursivement(dossier: DocumentFile, chemin: List<String>): List<FichierAVectoriser> {
        val resultat = mutableListOf<FichierAVectoriser>()
        dossier.listFiles().forEach { enfant ->
            val nom = enfant.name ?: return@forEach
            if (enfant.isDirectory) {
                resultat.addAll(parcourirRecursivement(enfant, chemin + nom))
            } else {
                val typeMime = enfant.type ?: "application/octet-stream"
                if (!typeMime.startsWith("video/")) {
                    resultat.add(FichierAVectoriser(enfant.uri, nom, chemin, typeMime, enfant.length()))
                }
            }
        }
        return resultat
    }

    fun renommer(elementUri: Uri, nouveauNom: String): Boolean {
        val doc = DocumentFile.fromSingleUri(context, elementUri) ?: return false
        return doc.renameTo(nouveauNom)
    }

    fun supprimer(elementUri: Uri): Boolean {
        val doc = DocumentFile.fromSingleUri(context, elementUri) ?: return false
        return doc.delete()
    }

    /**
     * Deplace un document vers un autre dossier PARENT, a l'interieur du meme
     * arbre designe (DocumentsContract.moveDocument ne fonctionne qu'au sein
     * d'un meme fournisseur de documents). anciensParentUri et nouveauParentUri
     * sont les Uri des dossiers source/destination (pas du document lui-meme).
     */
    fun deplacer(elementUri: Uri, ancienParentUri: Uri, nouveauParentUri: Uri): Boolean {
        return try {
            val resolver: ContentResolver = context.contentResolver
            val ancienParentDocId = DocumentsContract.getTreeDocumentId(ancienParentUri)
                .let { DocumentsContract.buildDocumentUriUsingTree(ancienParentUri, it) }
            val nouveauParentDocId = DocumentsContract.getTreeDocumentId(nouveauParentUri)
                .let { DocumentsContract.buildDocumentUriUsingTree(nouveauParentUri, it) }
            DocumentsContract.moveDocument(resolver, elementUri, ancienParentDocId, nouveauParentDocId) != null
        } catch (e: Exception) {
            false
        }
    }
}

File: ios/App/App/dossiers/DossiersDesignesRepository.swift
// Cree le 23/08/2026, Bourama : Lot 2 Partie 3 (app mobile), fichiers/dossiers designes, iOS.
//
// UIDocumentPickerViewController (mode dossier) + security-scoped bookmarks
// pour la persistance (voir 02-fichiers-dossiers.md). Contrairement a
// Android (une permission URI persistante suffit), iOS exige de
// "start/stopAccessingSecurityScopedResource()" a CHAQUE usage -- ce n'est
// pas une permission qu'on prend une fois pour toutes cote systeme, c'est
// une session d'acces a ouvrir/fermer a chaque operation. D'ou le wrapper
// `avecAcces(...)` ci-dessous, utilise par toutes les operations.
//
// Bookmarks stockes dans UserDefaults (Data en base64 implicite via
// UserDefaults). Pas besoin d'entitlement special pour ceci -- contrairement
// a Screen Time (voir README.md), UIDocumentPickerViewController fonctionne
// sur un compte Apple Developer standard.
import Foundation
import UniformTypeIdentifiers

private let CLE_BOOKMARKS = "dossiers_designes_bookmarks"

struct DossierDesigne: Identifiable {
    let id: String // bookmark encode en base64, sert aussi de cle de retrait
    let url: URL
}

struct ElementDossier: Identifiable {
    let id: String
    let url: URL
    let nom: String
    let estDossier: Bool
    let tailleOctets: Int64
}

// Ajoute le 30/08/2026 (correctif Claude chat) : voir DossiersDesignesRepository.kt
// cote Android pour le role exact, meme structure des deux cotes.
struct ContenuFichier {
    let contenuBase64: String
    let typeMime: String
    let nomFichier: String
    let tailleOctets: Int64
}

// Ajoute le 04/09/2026 (vectorisation en masse des dossiers designes,
// voir DossiersDesignesRepository.kt cote Android pour le meme role) :
// un fichier trouve en parcourant recursivement un dossier designe, avec
// son chemin de sous-dossiers depuis la racine (jamais le nom du dossier
// designe lui-meme, ni le nom du fichier).
struct FichierAVectoriser {
    let url: URL
    let nom: String
    let chemin: [String]
    let typeMime: String
    let tailleOctets: Int64
}

enum DossiersDesignesRepository {

    // MARK: - Liste des dossiers designes

    static func listerDossiersDesignes() -> [DossierDesigne] {
        let bookmarks = UserDefaults.standard.array(forKey: CLE_BOOKMARKS) as? [Data] ?? []
        return bookmarks.compactMap { bookmark in
            var estPerime = false
            guard let url = try? URL(
                resolvingBookmarkData: bookmark,
                options: [],
                relativeTo: nil,
                bookmarkDataIsStale: &estPerime
            ) else { return nil }
            // Un bookmark perime (dossier deplace/renomme cote systeme) est quand
            // meme retourne ici -- il sera simplement retire au prochain retrait
            // manuel si l'acces echoue reellement. Ne pas le faire disparaitre
            // silencieusement de la liste, l'etudiant doit pouvoir le voir et le retirer.
            return DossierDesigne(id: bookmark.base64EncodedString(), url: url)
        }
    }

    static func ajouterDossierDesigne(url: URL) {
        guard url.startAccessingSecurityScopedResource() else { return }
        defer { url.stopAccessingSecurityScopedResource() }
        guard let bookmark = try? url.bookmarkData(options: [], includingResourceValuesForKeys: nil, relativeTo: nil) else { return }
        var bookmarks = UserDefaults.standard.array(forKey: CLE_BOOKMARKS) as? [Data] ?? []
        bookmarks.append(bookmark)
        UserDefaults.standard.set(bookmarks, forKey: CLE_BOOKMARKS)
    }

    static func retirerDossierDesigne(id: String) {
        var bookmarks = UserDefaults.standard.array(forKey: CLE_BOOKMARKS) as? [Data] ?? []
        bookmarks.removeAll { $0.base64EncodedString() == id }
        UserDefaults.standard.set(bookmarks, forKey: CLE_BOOKMARKS)
    }

    // MARK: - Acces securise (a ouvrir/fermer a chaque operation, voir en-tete)

    private static func avecAcces<T>(_ url: URL, _ operation: () throws -> T) -> T? {
        guard url.startAccessingSecurityScopedResource() else { return nil }
        defer { url.stopAccessingSecurityScopedResource() }
        return try? operation()
    }

    // MARK: - CRUD

    static func listerContenu(_ dossier: URL) -> [ElementDossier] {
        avecAcces(dossier) {
            let fm = FileManager.default
            let enfants = try fm.contentsOfDirectory(
                at: dossier,
                includingPropertiesForKeys: [.isDirectoryKey, .fileSizeKey],
                options: [.skipsHiddenFiles]
            )
            return enfants.map { url in
                let valeurs = try? url.resourceValues(forKeys: [.isDirectoryKey, .fileSizeKey])
                let estDossier = valeurs?.isDirectory ?? false
                let taille = Int64(valeurs?.fileSize ?? 0)
                return ElementDossier(id: url.path, url: url, nom: url.lastPathComponent, estDossier: estDossier, tailleOctets: taille)
            }.sorted {
                if $0.estDossier != $1.estDossier { return $0.estDossier && !$1.estDossier }
                return $0.nom.localizedCaseInsensitiveCompare($1.nom) == .orderedAscending
            }
        } ?? []
    }

    static func creerSousDossier(dansParent parent: URL, nom: String) -> Bool {
        avecAcces(parent) {
            let destination = parent.appendingPathComponent(nom, isDirectory: true)
            try FileManager.default.createDirectory(at: destination, withIntermediateDirectories: false)
        } != nil
    }

    // Modifie le 05/09/2026, Bourama : ajoute "typeMime" (optionnel, defaut
    // nil pour ne rien changer au chemin d'appel existant depuis
    // DossiersPlugin.creerFichier, cote UI web/JS). Corrige un ecart avec
    // Android : la-bas, DocumentFile.createFile(typeMime, nom) via SAF
    // choisit lui-meme l'extension a partir du type MIME quand "nom" n'en a
    // pas deja une. Cote iOS, FileManager n'a aucune notion de type MIME,
    // le type d'un fichier est entierement determine par l'extension dans
    // son nom -- sans ce correctif, un "nom" recu sans extension (l'agent
    // comptant sur "type_mime" pour l'implicite) donnait un fichier sans
    // extension sur iPhone alors qu'Android l'aurait complete. Voir
    // nomAvecExtension ci-dessous.
    static func creerFichier(dansParent parent: URL, nom: String, typeMime: String? = nil) -> Bool {
        avecAcces(parent) {
            let nomFinal = Self.nomAvecExtension(nom, typeMime: typeMime)
            let destination = parent.appendingPathComponent(nomFinal, isDirectory: false)
            guard FileManager.default.createFile(atPath: destination.path, contents: Data()) else {
                throw NSError(domain: "DossiersDesignesRepository", code: 1)
            }
        } != nil
    }

    // Si "nom" a deja une extension, ne jamais la toucher (meme principe
    // que le reste du projet : ne jamais deviner par-dessus ce qui est deja
    // explicite). Sinon, en deduire une a partir du type MIME recu, si
    // fourni et reconnu -- sinon "nom" reste tel quel (aucune extension
    // ajoutee), meme comportement qu'avant ce correctif.
    private static func nomAvecExtension(_ nom: String, typeMime: String?) -> String {
        guard (nom as NSString).pathExtension.isEmpty,
              let typeMime = typeMime,
              let ext = UTType(mimeType: typeMime)?.preferredFilenameExtension else {
            return nom
        }
        return "\(nom).\(ext)"
    }

    static func renommer(_ element: URL, nouveauNom: String) -> Bool {
        avecAcces(element.deletingLastPathComponent()) {
            let destination = element.deletingLastPathComponent().appendingPathComponent(nouveauNom)
            try FileManager.default.moveItem(at: element, to: destination)
        } != nil
    }

    static func supprimer(_ element: URL) -> Bool {
        avecAcces(element.deletingLastPathComponent()) {
            try FileManager.default.removeItem(at: element)
        } != nil
    }

    static func deplacer(_ element: URL, versParent nouveauParent: URL) -> Bool {
        avecAcces(nouveauParent) {
            let destination = nouveauParent.appendingPathComponent(element.lastPathComponent)
            try FileManager.default.moveItem(at: element, to: destination)
        } != nil
    }

    // MARK: - Vectorisation en masse (04/09/2026)

    // Parcourt recursivement dossier (et sous-dossiers), renvoie chaque
    // FICHIER (jamais les dossiers), video exclue -- meme role que
    // listerRecursif cote Android (DossiersDesignesRepository.kt), voir
    // ce fichier pour le contexte complet. Contrairement aux autres
    // fonctions de ce fichier, NE gere PAS elle-meme l'acces
    // security-scoped : l'appelant (DossiersPlugin, voir
    // demarrerVectorisationDossier) doit deja avoir ouvert
    // startAccessingSecurityScopedResource() sur la racine et le garder
    // ouvert pendant toute la duree du transfert (listage + lecture +
    // upload de chaque fichier) -- rouvrir/refermer l'acces a chaque
    // fichier serait inutilement couteux et fragile.
    static func listerRecursif(_ dossier: URL, chemin: [String] = []) -> [FichierAVectoriser] {
        let fm = FileManager.default
        guard let enfants = try? fm.contentsOfDirectory(
            at: dossier,
            includingPropertiesForKeys: [.isDirectoryKey, .fileSizeKey],
            options: [.skipsHiddenFiles]
        ) else { return [] }
        var resultat: [FichierAVectoriser] = []
        for url in enfants {
            let valeurs = try? url.resourceValues(forKeys: [.isDirectoryKey, .fileSizeKey])
            let estDossier = valeurs?.isDirectory ?? false
            if estDossier {
                resultat.append(contentsOf: listerRecursif(url, chemin: chemin + [url.lastPathComponent]))
            } else {
                let typeMime = UTType(filenameExtension: url.pathExtension)?.preferredMIMEType
                    ?? "application/octet-stream"
                if !typeMime.hasPrefix("video/") {
                    let taille = Int64(valeurs?.fileSize ?? 0)
                    resultat.append(FichierAVectoriser(url: url, nom: url.lastPathComponent, chemin: chemin, typeMime: typeMime, tailleOctets: taille))
                }
            }
        }
        return resultat
    }

    // Ajoute le 30/08/2026 (correctif Claude chat, meme role que la version
    // Android, voir son en-tete) : contenu brut d'un fichier deja repere,
    // encode en base64 pour transiter par le canal temps reel WebSocket.
    static func lireFichier(_ element: URL) -> ContenuFichier? {
        avecAcces(element.deletingLastPathComponent()) {
            let data = try Data(contentsOf: element)
            let typeMime = UTType(filenameExtension: element.pathExtension)?.preferredMIMEType
                ?? "application/octet-stream"
            return ContenuFichier(
                contenuBase64: data.base64EncodedString(),
                typeMime: typeMime,
                nomFichier: element.lastPathComponent,
                tailleOctets: Int64(data.count)
            )
        }
    }
}
